rfcs/2020-04-06-1858-automatically-adjust-limits.md

```diff
@@ -0,0 +1,164 @@
+# RFC 1858 - 2020-04-06 - Automatically Adjust Limits
+
+This RFC proposes a new scheme for rate limiting requests to external
+services in order to maximize the sustained transmission rate over
+varying conditions.
+
+## Motivation
+
+Vector users commonly run into the problem of internal service rate
+limiting. This is not an external service limiting receiving data from
+us but our own rate limiting based on the `request`
+parameters. Alternatively, users can run into the opposite problem of
+overwhelming a downstream service to the point where it becomes
+unresponsive and starts queueing requests. Instead, Vector should be
+automatically rate limiting its requests to maximally fill the service's
+capacity without overwhelming it and causing additional problems.
+
+Most sinks in Vector have their request structure managed by the [tower
+crate](https://github.com/tower-rs/tower).  This service builder allows
+for setting how requests are sent to remote sinks. In particular, Vector
+fixes the number of requests that may be simultaneously in flight (AKA
+the concurrency limit), the maximum rate at which requests may be sent,
+expressed in terms of the number of requests over some time interval
+(AKA the rate limit number and duration).
+
+Many of these parameters _must_ be adjusted by the Vector administrator
+to maximize the throughput for each service. For high volume sites, this
+can require considerable trial and error experimentation before a
+satisfactory set of parameters can be achieved. More importantly,
+changes in the service parameters, whether in terms of available
+processing power, bandwidth, or changes to latency, such as is caused by
+link congestion, or the number of other agents delivering to the same
+sink, can cause the same settings to now impede utilization rather than
+improve it.
+
+Since many of these factors that affect delivery rate are not fixed
+values but instead will vary considerably throughout the life of a
+service, it is impossible to choose the "best" parameters that will fit
+all these conditions. Instead, Vector should adopt an approach that
+allows for dynamic adjustment of the underlying parameters based on the
+current conditions.
+
+When service limits are reached, Vector will experience a number of
+undesirable phenomenon, notably request latency, timeouts, and
+deferrals. These all decrease the overall flow rate while actually
+increasing actual bandwidth usage.
+
+## Guide-level Proposal
+
+There are two levels of controls we have in play—static and dynamic. The
+existing controls are static, and best describe service limits, such as
+maximum allowed request rates. What is needed is a set of dynamic
+controls that adapt to the underlying conditions and scale the service
+utilization on the fly.
+
+Since the controls under consideration are dependant on some form of
+queueing, the controls will be inserted at the same level as
+`TowerRequestSettings`. The existing rate limit controls will remain in
```

Comment:
I like this. I assume users could also disable the automatic adjustment to preserve the current static behavior?

Reply:
Both you and @lukesteensen commented to this effect, so it seems desirable.